### PR TITLE
[jhw] Dynamo table API (Create, Update, Describe, Delete, and some of the TimeToLive functions)

### PR DESCRIPTION
**Pull request by jweissBHG**
Not sure if this is 100% ready to go, there are a couple things I wanted to propose and see how you feel about. For my use-case we are managing tables through C#, so I needed this functionality (I have tested it in a separate project).

1) I added a `IConvertibleFromJson` that allowed me to make extension methods `GetObject<T>(this JsonElement)` and `GetObjectArray<T>(this JsonElement)` for quicker writing of the deserialization functions. If you're comfortable with this convention, I feel like Carbon.Json could use a `IConvertibleToJson` with a `JsonObject ToJson()` function on it, to make serialization quicker to write as well. If you don't like, I'm very much open to alternative solutions. Just didn't want to be too verbose in the `FromJsonElement()` functions. This also makes the `DynamoDbClient.cs` code less verbose per function, as we can just do `return await SendAndReadJsonAsync<T>(httpRequest).ConfigureAwait(false);` with the type of the response object

2) Is it odd to be using `System.Text.Json` for deserializing, but `Carbon.Json` for serializing the Dynamo requests? Is the intention to move everything over to `System.Text.Json`?

3) I'm not really sure the best way to add unit tests for these SDK calls. I know that they work now enough for my use-case, but I can't be sure all the odds and ends will actually be accepted by the AWS API (like table stream configuration, or SSE configuration, stuff I'm not currently using). Is it reasonable to just fix issues as they are found, rather than trying to get full unit tests for every combination of field?

### CODE REVIEW

**Comment by iamcarbon** [comment]
I'll give this a more through review tomorrow. But initial comments on two and three. 

2. System.Text.Json is much faster, eliminates many allocations, and the extra UTF16-UTF8 conversions. I'm aiming to eliminate the final uses of Carbon.Json before v1. I've added an initial round of functions to ease the migration, and up for adding more until we can finally remove it. 

3. If API is sound and unlocks a use case, I think it's reasonable to get any new functionality in and fix any issues until the library reaches 1.0.

**Comment by iamcarbon** [comment]
OK, took a second review. Kill the ToJson() pattern and lets serialize these directly using System.Text.Json. We'll add an override to the base function that can serialize these directly.

File: src/Amazon.DynamoDb/Actions/UpdateTimeToLiveRequest.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        public UpdateTimeToLiveRequest(string tableName, string attributeName, bool enabled)
+        {
+            TableName = tableName ?? throw new ArgumentNullException(nameof(tableName));
+            TimeToLiveSpecification = new TimeToLiveSpecification(attributeName, enabled);
+        }
+
+        public string TableName { get; }
+        public TimeToLiveSpecification TimeToLiveSpecification { get; }
+
+        public JsonObject ToJson()
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Actions/UpdateTimeToLiveRequest.cs]: Kill & serialize directly using System.Text.Json

File: src/Amazon.DynamoDb/DynamoDbClient.cs [hunk 2]
```diff
@@ -61,17 +62,26 @@ public async Task<DeleteItemResult> DeleteItemAsync(DeleteItemRequest request)
             return DeleteItemResult.FromJsonElement(json);
         }
 
-        /*
-        public void DeleteTable()
+        public async Task<DeleteTableResult> DeleteTableAsync(string tableName)
+        {
+            var httpRequest = Setup("DeleteTable", new TableRequest(tableName).ToJson());
+
+            return await SendAndReadJsonAsync<DeleteTableResult>(httpRequest).ConfigureAwait(false);
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/DynamoDbClient.cs]: Create a private SendJsonAndReadJsonAsync<TSource, TTarget> overload that serializes the request using System.Text.Json

File: src/Amazon.DynamoDb/Responses/ListTablesResult.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace Amazon.DynamoDb
+{
+    public sealed class ListTablesResult
+    {
+        public string LastEvaluatedTableName { get; set; }
+        public bool HasMoreTables => LastEvaluatedTableName != null;
+        public string[] TableNames { get; set; }
+
+        public static ListTablesResult FromJsonElement(JsonElement el)
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Responses/ListTablesResult.cs:14]: Kill and Serialize directly using System.Text.Json.JsonSerializer

File: src/Amazon.DynamoDb/Responses/UpdateTimeToLiveResult.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+using System;
+using System.Text.Json;
+
+namespace Amazon.DynamoDb
+{
+    public class UpdateTimeToLiveResult : IConvertibleFromJson
+    {
+        public TimeToLiveSpecification TimeToLiveSpecification { get; set; }
+
+        public void FillField(JsonProperty property)
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Responses/UpdateTimeToLiveResult.cs]: Kill and Serialize directly using System.Text.Json.JsonSerializer

File: src/Amazon.DynamoDb/Models/TimeToLiveSpecification.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        public JsonObject ToJson()
+        {
+            return new JsonObject
+            {
+                { "AttributeName", AttributeName },
+                { "Enabled", Enabled },
+            };
+        }
+
+        public static TimeToLiveSpecification FromJson(JsonNode json)
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Models/TimeToLiveSpecification.cs]: Kill and deserialize directly using System.Text.Json.JsonSerializer

File: src/Amazon.DynamoDb/Models/TimeToLiveDescription.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+using System.Text.Json;
+
+namespace Amazon.DynamoDb
+{
+    public class TimeToLiveDescription : IConvertibleFromJson
+    {
+        public string? AttributeName { get; set; }
+        public TimeToLiveStatus TimeToLiveStatus { get; set; }
+
+        public void FillField(JsonProperty property)
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Models/TimeToLiveDescription.cs:14]: Kill FillField

File: src/Amazon.DynamoDb/Models/TableStatus.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Amazon.DynamoDb
+{
+    public enum TableStatus : byte
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Models/TableStatus.cs:6]: Remove : byte

File: src/Amazon.DynamoDb/Models/TableDescription.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        public RestoreSummary[]? RestoreSummary { get; set;}
+        public SSEDescription? SSEDescription { get; set; }
+        public StreamSpecification? StreamSpecification { get; set; }
+        public string? TableArn { get; set; }
+        public string? TableId { get; set; }
+        public string? TableName { get; set; }
+        public long TableSizeBytes { get; set; }
+        public TableStatus TableStatus { get; set; }
+
+        public void FillField(JsonProperty p)
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Models/TableDescription.cs]: Kill and deserialize directly using System.Text.Json.JsonSerializer

File: src/Amazon.DynamoDb/Models/TableStatus.cs (hunk 7 above)
Comment (iamcarbon) [reply src/Amazon.DynamoDb/Models/TableStatus.cs:6]: Add [JsonConverter(typeof(JsonStringEnumConverter))]

File: src/Amazon.DynamoDb/Models/TableStatus.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        CREATING,
+        UPDATING,
+        DELETING,
+        ACTIVE,
+        INACCESSIBLE_ENCRYPTION_CREDENTIALS,
+        ARCHIVING,
+        ARCHIVED,
+    };
+
+    public static class TableStatusExtensions
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Models/TableStatus.cs]: Kill this class

File: src/Amazon.DynamoDb/Models/TimeToLiveStatus.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+{
+    public enum TimeToLiveStatus : byte
+    {
+        ENABLING,
+        DISABLING,
+        ENABLED,
+        DISABLED,
+    };
+
+    public static class TimeToLiveStatusExtensions
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Models/TimeToLiveStatus.cs]: Kill TimeToLiveStatusExtensions

File: src/Amazon.DynamoDb/Models/TimeToLiveStatus.cs [hunk 11]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Amazon.DynamoDb
+{
+    public enum TimeToLiveStatus : byte
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Models/TimeToLiveStatus.cs:6]: Remove byte and add [JsonConverter(typeof(StringEnumJsonConverter))]

File: src/Amazon.DynamoDb/Models/UpdateGlobalSecondaryIndexAction.cs [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        public UpdateGlobalSecondaryIndexAction(string indexName, ProvisionedThroughput provisionedThroughput)
+        {
+            IndexName = indexName ?? throw new ArgumentNullException(nameof(indexName));
+            ProvisionedThroughput = provisionedThroughput ?? throw new ArgumentNullException(nameof(provisionedThroughput));
+        }
+
+        public string IndexName { get; }
+        public ProvisionedThroughput ProvisionedThroughput { get; set; }
+
+        public JsonObject ToJson()
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Models/UpdateGlobalSecondaryIndexAction.cs:20]: Kill JsonObject ToJson() and Serialize directly using System.Text.Json

File: src/Amazon.DynamoDb/Responses/CreateTableResult.cs [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+namespace Amazon.DynamoDb
+{
+    public sealed class CreateTableResult : IConvertibleFromJson
+    {
+        public CreateTableResult() { }
+
+        public TableDescription TableDescription { get; set; }
+
+        public void FillField(JsonProperty property)
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Responses/CreateTableResult.cs]: Kill and Serialize directly using System.Text.Json.JsonSerializer

File: src/Amazon.DynamoDb/Responses/DescribeTableResult.cs [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+namespace Amazon.DynamoDb
+{
+    public sealed class DescribeTableResult : IConvertibleFromJson
+    {
+        public DescribeTableResult() { }
+
+        public TableDescription TableDescription { get; set; }
+
+        public void FillField(JsonProperty property)
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Responses/DescribeTableResult.cs]: Kill and Serialize directly using System.Text.Json.JsonSerializer

File: src/Amazon.DynamoDb/Responses/UpdateTableResult.cs [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+
+namespace Amazon.DynamoDb
+{
+    public sealed class UpdateTableResult : IConvertibleFromJson
+    {
+        public UpdateTableResult() { }
+
+        public TableDescription TableDescription { get; set; }
+
+        public void FillField(JsonProperty property)
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Responses/UpdateTableResult.cs]: Kill and Serialize directly using System.Text.Json.JsonSerializer

File: src/Amazon.DynamoDb/Models/ArchivalSummary.cs [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Amazon.DynamoDb
+{
+    public class ArchivalSummary : IConvertibleFromJson
+    {
+        public string? ArchivalBackupArn { get; set; }
+        public string? ArchivalDateTime { get; set; }
+        public string? ArchivalReason { get; set; }
+
+        public void FillField(JsonProperty property)
```
Comment (iamcarbon) [inline src/Amazon.DynamoDb/Models/ArchivalSummary.cs]: Kill and Serialize directly using System.Text.Json.JsonSerializer

**Comment by iamcarbon** [comment]
These new models should be serialized directly System.Text.Json.Serializer.Serialize(). Otherwise, all the new APIs look good!

Let me know if you need help on converting or extending the base client.

**Comment by iamcarbon** [comment]
@jweissBHG Let me know if you think we should merge as-is, and do another update to convert over too.

**Comment by jweissBHG** [comment]
@iamcarbon sorry for letting this sit, I've been working on other stuff and letting this version of Carbon SDK soak in some user testing for a couple weeks (I had everything I needed to turn it on in our project). The good news is everything is working great, so I think we will continue to use your SDK over the official one. I'm gonna clean up this PR and move on to adding support for DAX and dynamo transactions

**Comment by jweissBHG** [comment]
Okay I've refactored to just use System.Text.Json for the serialization and deserialization. Testing this now in my codebase. I haven't changed any of the other pre-existing calls to this way, just the new Table functions

**Comment by iamcarbon** [comment]
@jweissBHG This looks good from my side. Let me know how your testing goes!

**Comment by jweissBHG** [comment]
First run failed due to DateTimeOffset serialization, I added a custom serializer to convert from the Double format that AWS API uses. This time tests passed (I have a script that creates tables, describes them, updates TTLs, and deletes them)

**Comment by jweissBHG** [comment]
Good to go now from my perspective

**Pull request merged**